planning/MartenV4.md

```diff
@@ -0,0 +1,114 @@
+# Marten v4 Ideas
+
+## Pulling it Off
+
+We've got the typical problems of needing to address incoming pull requests and bug issues in master while probably needing to have a long lived branch for *v4*. 
+
+Here are some thoughts:
+
+* First do a bug sweep 3.12 release to address as many of the tactical problems as we can before branching to v4?
+* Start with the unit testing improvements as a way to speed up the build before we dive into much more of this? 
+* Possibly do a series of v4, then v5 releases to do this in smaller chunks? We've mostly said do the event store as v4, then Linq improvements as v5
+* Do the Event Store v4 work in a separate project built as an add on from the very beginning, but leave the existing event store in place. That would enable us to do a lot of work *and* mostly be able to keep that work in master so we don't have long-lived branch problems
+* Damn the torpedos and full speed ahead!
+
+## Miscellaneous Ideas
+
+* Look at some kind of object pooling for the `DocumentSession` / `QuerySession` objects?
+* Ditch the document by document type schema configuration. Do that, and I think we open the door for multi-tenancy by schema
+* Eliminate `ManagedConnection` altogether. I think it results in unnecessary object allocations and it's causing more harm that help as it's been extended over time
```

Comment:
Do we plan to resign in general from the connection pool management?

Reply:
Connection pool management is Npgsql, not us. `ManagedConnection` is just a convenience to simplify ADO.Net mechanics for developers, but it's not super-efficient at runtime because of the extra object allocations with the Lambdas.

Reply:
That's fine however it would be good to check how the EF is doing the connection pooling. Especially if we'd like to do cross schema or later cross database tenancy.

Reply:
I wouldn't expect that EF Core does its own connection pooling, or do you just mean how it deals with connection lifecycle?

Reply:
Do you just mean the connection lifecycle management? Because EF wouldn't be doing the actual connection pooling, that'd be deeper in the ADO.Net provider level

Reply:
My bad. You're right. They're not doing connection pooling but the connection lifecycle management (so probably similar as we do).

Reply:
I added more comments. I don't think we can get away with eliminating `ManagedConnection` easily, but I want some changes to stop the inefficient Lambda usage.